zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java

```diff
@@ -220,6 +224,24 @@ public int getSslHandshakeTimeoutMillis() {
     }
 
     public SSLContext createSSLContext(ZKConfig config) throws SSLContextException {
+        if (config.getProperty(sslClientContextProperty) != null) {
+            LOG.debug("Loading SSLContext from property '" + sslClientContextProperty + "'");
+            String sslClientContextClass = config.getProperty(sslClientContextProperty);
+            try {
+                Class<?> sslContextClass = Class.forName(sslClientContextClass);
+                ZKClientSSLContext sslContext = (ZKClientSSLContext) sslContextClass.getConstructor().newInstance();
```

Comment:
I would use `java.util.function.Supplier<SSLContext>` instead of your own `ZKClientSSLContext` type here

Reply:
I'm not exactly sure what you mean here. Supplier is a functional interface, but we're initialising it from a property. Do you mean that users should implement the Supplier interface in the class that they pass in? That wouldn't really seem to match the intent of functional interfaces - which would generally be passed as a lambda expression or method reference.

Reply:
Functional interfaces allow lambdas and method references but I don't think they are restricted to it. But anyway, I don't feel too strongly about it. I don't think `Client` should be in the class name (since this works for both `QuorumX509Util` and `ClientX509Util`) so if you stick with your own class, I would call the interface something like `SSLContextSupplier`.

Reply:
I can use the Supplier interface - I just wanted to double check before updating the PR. Haven't really seen it implemented, but that's something that's a bit unique to this use-case. It is just an interface at the end of the day.

Renaming the property is grand as well - the name made sense back before ClientX509Util for my specific use-case, but given the changes, renaming it makes more sense.

Just checking - we get an unchecked cast warning when instantiating the supplier and casting it to Supplier<SSLContext> - should I leave that warning, or suppress that warning?

Reply:
do you still get the warning if you do an explicit `instanceof` check first, and throw an exception if the check fails?

Reply:
The problem is type-erasure - so we can verify that it's an instanceof Supplier, just not Supplier<SSLContext>. We'll get a ClassCastException either way though - which is already handled in the try/catch.